swupd/helpers_test.go

```diff
@@ -361,9 +360,165 @@ func (fs *testFileSystem) cp(src, dst string) {
 }
 
 func (fs *testFileSystem) rm(subpath string) {
+	fs.t.Helper()
 	path := filepath.Join(fs.Dir, subpath)
 	err := os.RemoveAll(path)
 	if err != nil {
 		fs.t.Fatalf("error removing %s: %s", subpath, err)
 	}
 }
+
+func (fs *testFileSystem) mkdir(subpath string) {
+	fs.t.Helper()
+	path := filepath.Join(fs.Dir, subpath)
+	err := os.MkdirAll(path, 0755)
+	if err != nil {
+		fs.t.Fatalf("error creating directory %s: %s", subpath, err)
+	}
+}
+
+func (fs *testFileSystem) exists(subpath string) bool {
+	fs.t.Helper()
+	path := filepath.Join(fs.Dir, subpath)
+	_, err := os.Stat(path)
+	switch {
+	case err == nil:
+		return true
+	case os.IsNotExist(err):
+		return false
+	default:
+		fs.t.Fatalf("error checking if %s exists: %s", subpath, err)
+		return false
+	}
+}
+
+func (fs *testFileSystem) checkContains(subpath, sub string) {
+	fs.t.Helper()
+	path := filepath.Join(fs.Dir, subpath)
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		fs.t.Errorf("couldn't open %s to check its contents: %s", subpath, err)
+	}
+	if !bytes.Contains(contents, []byte(sub)) {
+		fs.t.Errorf("%s did not contain expected %q", subpath, sub)
+	}
+}
+
+func (fs *testFileSystem) checkNotContains(subpath, sub string) {
+	fs.t.Helper()
+	path := filepath.Join(fs.Dir, subpath)
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		fs.t.Errorf("couldn't open %s to check its contents: %s", subpath, err)
+	}
+	if bytes.Contains(contents, []byte(sub)) {
+		fs.t.Errorf("%s did not contain expected %q", subpath, sub)
+	}
+}
+
+// testSwupd is a struct that keeps track of an update content state and a testing.T. It
+// can be used to perform repository construction and operations and handle unexpected
+// errors with t.Fatal. It does embed a testFileSystem, so all the filesystem operations
+// are also available.
+//
+// It is to be used in places where the swupd operations (create manifests, create packs)
+// are expected to return without errors and the subject of the test is their product. In
+// case errors of those operations are to be tested, use the data from the struct but not
+// the helper functions.
+//
+// Simple usage looks like
+//
+// func TestMyTest(t *testing.T) {
+// 	ts := newTestSwupd(t, "my-test-")
+// 	defer ts.cleanup()
+//
+//	ts.Bundles = []string{"bundle"}
+//      ts.write("image/10/bundle/file", "content")
+//      ts.createManifests(10)
+//
+//      // ...
+// }
+//
+// For tests that require only filesystem operations, prefer testFileSystem.
+type testSwupd struct {
+	*testFileSystem
+
+	Bundles    []string
+	MinVersion uint32
+	Format     uint
+}
+
+func newTestSwupd(t *testing.T, prefix string) *testSwupd {
+	fs := newTestFileSystem(t, prefix)
+	defer func() {
+		if t.Failed() {
```

Comment:
Do you mean `if !t.Failed()`? This looks like it is cleaning up only if the test failed.

Reply:
The code is correct, but due to similarity how defer cleanup is used elsewhere might be confusing. If when creating a testSwupd we have a failure, cleanup any existing directory. Otherwise let it live because the caller will call cleanup themselves.

I'll add a comment there.

Reply:
Note that currently fs.cleanup won't really cleanup (because it was a failure) but will still show the helpful message with the path of the temporary directory.

Reply:
Nice, thanks.